server/src/main/java/org/opensearch/common/blobstore/transfer/RemoteTransferContainer.java

```diff
@@ -0,0 +1,180 @@
+        this.contentLength = contentLength;
+        this.failTransferIfFileExists = failTransferIfFileExists;
+        this.writePriority = writePriority;
+        this.offsetRangeInputStreamSupplier = offsetRangeInputStreamSupplier;
+    }
+
+    /**
+     * @return The {@link  WriteContext} for the current upload
+     */
+    public WriteContext createWriteContext() {
+        return new WriteContext(
+            remoteFileName,
+            this::supplyStreamContext,
+            contentLength,
+            failTransferIfFileExists,
+            writePriority,
+            this::finalizeUpload
+        );
+    }
+
+    /**
+     * @param partSize Part sizes of all parts apart from the last one, which is determined internally
+     * @return The {@link StreamContext} object that will be used by the vendor plugin to retrieve streams
+     */
+    public StreamContext supplyStreamContext(long partSize) {
+        try {
+            return this.openMultipartStreams(partSize);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private StreamContext openMultipartStreams(long partSize) throws IOException {
+        if (inputStreams.get() != null) {
+            throw new IOException("Multi-part streams are already created.");
+        }
+
+        this.partSize = partSize;
+        this.lastPartSize = (contentLength % partSize) != 0 ? contentLength % partSize : partSize;
+        this.numberOfParts = (int) ((contentLength % partSize) == 0 ? contentLength / partSize : (contentLength / partSize) + 1);
+        ResettableCheckedInputStream[] streams = new ResettableCheckedInputStream[numberOfParts];
+        inputStreams.set(streams);
+
+        return new StreamContext(new StreamProvider(getTransferPartStreamSupplier(), partSize, lastPartSize, numberOfParts), numberOfParts);
+    }
+
+    private TransferPartStreamSupplier getTransferPartStreamSupplier() {
+        return ((partNo, size, position) -> {
+            assert inputStreams.get() != null : "expected inputStreams to be initialised";
+            return getMultipartStreamSupplier(partNo, size, position).get();
+        });
+    }
+
+    /**
+     * OffsetRangeInputStreamSupplier is used to get the offset based input streams at runtime
+     */
+    public interface OffsetRangeInputStreamSupplier {
+        OffsetRangeInputStream get(long size, long position) throws IOException;
+    }
+
+    interface LocalStreamSupplier<Stream> {
+        Stream get() throws IOException;
+    }
+
+    private LocalStreamSupplier<Stream> getMultipartStreamSupplier(final int streamIdx, final long size, final long position) {
+        return () -> {
+            try {
+                OffsetRangeInputStream offsetRangeInputStream = offsetRangeInputStreamSupplier.get(size, position);
+                ResettableCheckedInputStream checkedInputStream = new ResettableCheckedInputStream(offsetRangeInputStream, localFileName);
```

Comment:
can we make this optional ? So we will need this only in case of encrypted data , so a flag would be helpful here.

Reply:
I'm making a change where data integrity check will be delegated to the vendor plugin based on the support on the plugin end and whether we are uploading decorated streams or not f.e. in case of encrypted streams.